pretyping/pretyping.ml

```diff
@@ -1025,12 +1043,29 @@ struct
         let subs = Esubst.subs_cons (Vars.make_substituend ujval) subs in
         let body = Coercion.push_arg body ujval in
         let val_before_bidi = if bidi then val_before_bidi else body in
-        apply_rec env sigma (n+1) body (subs, c2) val_before_bidi candargs bidiargs rest
+        apply_rec env sigma (n+1) body (subs, c2) val_before_bidi candargs template bidiargs rest
     in
     let typ = (Esubst.subs_id 0, fj.uj_type) in
     let body = (Coercion.start_app_body sigma fj.uj_val) in
-    let sigma, resj, val_before_bidi, bidiargs =
-      apply_rec env sigma 0 body typ body candargs [] args
+    let template_params, template = match EConstr.kind sigma fj.uj_val with
+      | Ind (ind,u) | Construct ((ind,_),u)
+        when EInstance.is_empty u && Environ.template_polymorphic_ind ind !!env ->
+        let mib = Environ.lookup_mind (fst ind) !!env in
+        Option.map (fun t -> t.Declarations.template_param_levels, Univ.Level.Map.empty)
+          mib.mind_template,
+        mib.mind_template
+      | _ -> None, None
+    in
+    let sigma, resj, val_before_bidi, template_params, bidiargs =
+      apply_rec env sigma 0 body typ body candargs template_params [] args
+    in
+    let sigma = match template_params, template with
+      | None, None -> sigma
+      | Some _, None | None, Some _ -> assert false
+      | Some (_,bound), Some t ->
+        let bound = Univ.Level.Map.map (fun u -> TemplateUniv u) bound in
+        let csts = Inductive.instantiate_template_constraints bound t in
+        Evd.add_constraints sigma csts
     in
     let sigma, resj = refresh_template env sigma resj in
```

Comment:
Why not use the bound substitution to compute the result type rather than refreshing here?

Reply:
Because it's TemplateUniv only, so it wouldn't work when we want to end up in Prop.

Reply:
I guess this is to handle Prop which can't instantiate the fresh level variable, but we could use a fresh quality instead?

Reply:
As discussed offline we should be able to remove the refresh but still need the retype